// trezor-blockchain-link/lib/utils/deferred.js
'use strict';

exports.__esModule = true;

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

exports.create = create;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function create(id) {
    var _this = this;

    var localResolve = function localResolve() {};
    var localReject = function localReject() {};

    var promise = new Promise(function () {
        var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(resolve, reject) {
            return _regenerator2.default.wrap(function _callee$(_context) {
                while (1) {
                    switch (_context.prev = _context.next) {
                        case 0:
                            localResolve = resolve;
                            localReject = reject;

                        case 2:
                        case 'end':
                            return _context.stop();
                    }
                }
            }, _callee, _this);
        }));

        return function (_x, _x2) {
            return _ref.apply(this, arguments);
        };
    }());

    return {
        id: id,
        resolve: localResolve,
        reject: localReject,
        promise: promise
    };
}
